pkg/pool/linked_list.go

```diff
@@ -0,0 +1,100 @@
+package pool
+
+import "sync"
+
+func NewDoubleLinkedList[T any]() *DoubleLinkedList[T] {
+	return &DoubleLinkedList[T]{}
+}
+
+type Node[T any] struct {
+	prev  *Node[T]
+	next  *Node[T]
+	value T
+}
+
+func (n *Node[T]) Value() T {
+	return n.value
+}
+
+type DoubleLinkedList[T any] struct {
+	lock sync.Mutex
+	head *Node[T]
+	len  uint64
+}
+
+func (l *DoubleLinkedList[T]) Len() uint64 {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	return l.len
+}
+
+func (l *DoubleLinkedList[T]) Insert(key T) *Node[T] {
```

Comment:
Use a named return here and avoid locking for as long as possible, and avoid defer wherever you can (defer is slow)

```go
 func (l *DoubleLinkedList[T]) Insert(key T) (node *Node[T]) {
    n = &Node[T]{
	value: key,
	next:  l.head,
    }
    l.lock.Lock()
    if l.head != nil {
	l.head.prev = newNode
    }
    l.head = newNode
    l.len++
    l.lock.Unlock()
    return
}
```